include/pistache/async.h

```diff
@@ -1446,22 +1450,25 @@ namespace Async {
             };
 
             template<typename ValueType, typename Dummy = void>
-            struct WhenContinuation {
-
-                typedef std::shared_ptr<DataT<ValueType>> D;
+            struct WhenContinuation
+            {
+                using D = std::shared_ptr<DataT<ValueType>>;
 
                 WhenContinuation(const D& _data, size_t _index)
                     : data(_data)
                     , index(_index)
                 { }
 
-                void operator()(const ValueType& val) const {
+                void operator()(const ValueType& val) const
+                {
+                    std::lock_guard<std::mutex> guard(data->mtx);
+
                     if (data->rejected) return;
 
                     data->results[index] = val;
-                    data->resolved.fetch_add(1);
-
-                    if (data->resolved == data->total) {
+                    data->resolved++;
+                    if (data->resolved == data->total)
+                    {
                         data->resolve(data->results);
```

Comment:
Is it appropriate to call data->resolve() while holding this mutex?

Reply:
@dennisjenkins75 To be honest, I don't see any problems (for instance, dead locks).
In the `data->resolve(data->results);` line we hold mutex from `Data` object and just call `operator()(Arg&& arg)` from `Resolver` object. Do you see any problems? Maybe I don't understand something.

Reply:
I did not look at what data->resolve() did (sorry, was busy).  Was just asking a general question. I should have looked first.  Its ok.